test/validation/api/classification/odp_classification_common.c

```diff
@@ -93,97 +93,60 @@ int stop_pktio(odp_pktio_t pktio)
 
 int cls_pkt_set_seq(odp_packet_t pkt)
 {
-	static uint32_t seq;
+	uint32_t seqno_offset = 0, l3_offset = 0, l3_len = 0;
+	odph_ipv4hdr_t *ipv4hdr;
+	odph_ipv6hdr_t *ipv6hdr;
 	cls_test_packet_t data;
-	uint32_t offset;
-	odph_ipv4hdr_t *ip;
-	odph_tcphdr_t *tcp;
-	odph_udphdr_t *udp;
-	uint16_t port = 0;
-	uint32_t hlen = 0;
+	static uint32_t seq;
+	uint8_t *buf;
 	int status;
 
+	buf = odp_packet_data(pkt);
+
+	if (odp_packet_has_ipv4(pkt)) {
+		l3_offset = odp_packet_l3_offset(pkt);
+		ipv4hdr = (odph_ipv4hdr_t *)(buf + l3_offset);
+		l3_len = odp_be_to_cpu_16(ipv4hdr->tot_len);
+		seqno_offset = l3_offset + l3_len - sizeof(data);
```

Comment:
Why not simply: seqno_offset = odp_packet_len(pkt) - sizeof(data);

Then the pair of functions would work also with non-IP packets and with
segmented packets.

Reply:
On Octeon, there is a configuration (NIX_AF_SMQX_CFG[MINLEN]) which tells the hardware to add padding if the packet received is below the threshold. So, when the testcase creates a 50b packet (ex:UDP dport testcase)  and sends it, on the receive side, packet gets padded and becomes 60b. The receive side code then looks for the seqno at a different offset and quite a few tests fail because of this. Hence, the logic is made to rely on L3 lengths.

Reply:
Ok, but it would still be good to write the code so that it does not assume single-segment packets. Like this:

```
static uint32_t seqno_offset(odp_packet_t pkt)
{
        uint32_t l3_offset = odp_packet_l3_offset(pkt);
        int rc;
        uint16_t len = 0;

        CU_ASSERT_FATAL(l3_offset != ODP_PACKET_OFFSET_INVALID);

        if (odp_packet_has_ipv4(pkt)) {
                odph_ipv4hdr_t ip;

                rc = odp_packet_copy_to_mem(pkt, l3_offset, sizeof(ip), &ip);
                CU_ASSERT_FATAL(rc == 0);
                len = odp_be_to_cpu_16(ip.tot_len);
        } else if (odp_packet_has_ipv6(pkt)) {
                odph_ipv6hdr_t ip;

                rc = odp_packet_copy_to_mem(pkt, l3_offset, sizeof(ip), &ip);
                CU_ASSERT_FATAL(rc == 0);
                len = sizeof(ip) + odp_be_to_cpu_16(ip.payload_len);
        } else
                CU_FAIL_FATAL("Unexcpected packet type");

        return l3_offset + len - sizeof(cls_test_packet_t);
}
```